target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch

```diff
@@ -0,0 +1,280 @@
+-		.name = "enable_vlan",
+-		.description = "VLAN mode (1:enabled)",
+-		.max = 1,
+-		.id = MT7530_ATTR_ENABLE_VLAN,
+-		.get = mt7530_get_vlan_enable,
+-		.set = mt7530_set_vlan_enable,
+-	},
+-};
+-
+ static u64 get_mib_counter(struct mt7530_priv *priv, int i, int port)
+ {
+ 	unsigned int port_base;
+@@ -683,6 +761,111 @@ static int mt7621_sw_get_port_mib(struct
+ 	return 0;
+ }
+ 
++static u64 get_mib_counter_7620(struct mt7530_priv *priv, int i)
++{
++	u64 t;
++
++	t = mt7530_r32(priv, MT7620_MIB_COUNTER_BASE + mt7620_mibs[i].offset);
++	if (mt7620_mibs[i].size == 2) {
++		u64 hi;
++
++		hi = mt7530_r32(priv, MT7620_MIB_COUNTER_BASE + mt7620_mibs[i].offset + 4);
++		t |= hi << 32;
++	}
++
++	return t;
++}
++
++static u64 get_mib_counter_port_7620(struct mt7530_priv *priv, int i, int port)
++{
++	unsigned int port_base;
++	u64 t;
++
++	port_base = MT7620_MIB_COUNTER_BASE_PORT +
++		    MT7620_MIB_COUNTER_PORT_OFFSET * port;
++
++	t = mt7530_r32(priv, port_base + mt7620_port_mibs[i].offset);
++	if (mt7620_mibs[i].size == 2) {
++		u64 hi;
++
++		hi = mt7530_r32(priv, port_base + mt7620_port_mibs[i].offset + 4);
++		t |= hi << 32;
++	}
++
++	return t;
++}
++
++static int mt7530_sw_get_mib(struct switch_dev *dev,
++				  const struct switch_attr *attr,
++				  struct switch_val *val)
++{
++	static char buf[4096];
++	struct mt7530_priv *priv = container_of(dev, struct mt7530_priv, swdev);
++	int i, len = 0;
++
++	len += snprintf(buf + len, sizeof(buf) - len, "Switch MIB counters\n");
++
++	for (i = 0; i < sizeof(mt7620_mibs) / sizeof(*mt7620_mibs); ++i) {
++		u64 counter;
++		len += snprintf(buf + len, sizeof(buf) - len,
++				"%-11s: ", mt7620_mibs[i].name);
++		counter = get_mib_counter_7620(priv, i);
++		len += snprintf(buf + len, sizeof(buf) - len, "%llu\n",
++				counter);
++	}
++
++	val->value.s = buf;
++	val->len = len;
++	return 0;
++}
++
++
++static int mt7530_sw_get_port_mib(struct switch_dev *dev,
++				  const struct switch_attr *attr,
++				  struct switch_val *val)
++{
++	static char buf[4096];
++	struct mt7530_priv *priv = container_of(dev, struct mt7530_priv, swdev);
++	int i, len = 0;
++
++	if (val->port_vlan >= MT7530_NUM_PORTS)
++		return -EINVAL;
++
++	len += snprintf(buf + len, sizeof(buf) - len,
++			"Port %d MIB counters\n", val->port_vlan);
++
++	for (i = 0; i < sizeof(mt7620_port_mibs) / sizeof(*mt7620_port_mibs); ++i) {
++		u64 counter;
++		len += snprintf(buf + len, sizeof(buf) - len,
++				"%-11s: ", mt7620_port_mibs[i].name);
++		counter = get_mib_counter_port_7620(priv, i, val->port_vlan);
++		len += snprintf(buf + len, sizeof(buf) - len, "%llu\n",
++				counter);
++	}
++
++	val->value.s = buf;
++	val->len = len;
++	return 0;
++}
++
++static const struct switch_attr mt7530_global[] = {
++	{
++		.type = SWITCH_TYPE_INT,
++		.name = "enable_vlan",
++		.description = "VLAN mode (1:enabled)",
++		.max = 1,
++		.id = MT7530_ATTR_ENABLE_VLAN,
++		.get = mt7530_get_vlan_enable,
++		.set = mt7530_set_vlan_enable,
++	}, {
```

Comment:
some parts need refactoring, made relocation part of code for new methods to be visible for further usage.